### PR TITLE
[Core] Adding lifecycle operations to Parameters

### PR DESCRIPTION
**Pull request by msandre**
This proposal includes the following changes.

- Default constructor, which constructs an empty, valid json string.
- Move ctor and assignment.
- Reset and swap.
- Tests for the added functions.

Additionally, string parameters passed by `const string` are replaced by `const string&`.

### CODE REVIEW

File: kratos/includes/kratos_parameters.h [hunk 1]
```diff
@@ -151,7 +152,7 @@ class Parameters {
   using iterator = iterator_adaptor;
   using const_iterator = const_iterator_adaptor;
 
-  Parameters(const std::string json_string) {
+  Parameters(const std::string& json_string = "{}") {
```
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h:155]: mmm, passing a reference to a string used to be problematic from python (at least valgrind did not love that).

I don't know if this is changed with pybind

File: kratos/includes/kratos_parameters.h [hunk 2]
```diff
@@ -233,7 +246,7 @@ class Parameters {
   }
 
   //*******************************************************************************************************
-  Parameters GetValue(const std::string entry) {
+  Parameters GetValue(const std::string& entry) {
```
Comment (RiccardoRossi) [inline kratos/includes/kratos_parameters.h:249]: same comment

File: kratos/python/add_kratos_parameters_to_python.cpp [hunk 3]
```diff
@@ -63,7 +63,8 @@ void  AddKratosParametersToPython(pybind11::module& m)
 
 
     class_<Parameters, Parameters::Pointer >(m,"Parameters")
-    .def(init<const std::string>()) //init<rapidjson::Value& >())
+    .def(init<>())
+    .def(init<const std::string&>()) //init<rapidjson::Value& >())
```
Comment (RiccardoRossi) [inline kratos/python/add_kratos_parameters_to_python.cpp]: ok, just take out the commented code

**Review by RiccardoRossi — DISMISSED** [review]

File: kratos/includes/kratos_parameters.h (hunk 1 above)
Comment (msandre) [reply kratos/includes/kratos_parameters.h:155]: It's done in the pybind examples, so it should be ok (https://pybind11.readthedocs.io/en/stable/classes.html?highlight=string%26)

File: kratos/python/add_kratos_parameters_to_python.cpp (hunk 3 above)
Comment (msandre) [reply kratos/python/add_kratos_parameters_to_python.cpp]: Done

**Review by RiccardoRossi — APPROVED** [review]
ok, i am approving, however if you may try to run the test_parameters.py with valgrind to see if it is clean

**Comment by msandre** [comment]
Tried valgrind. Looks ok. Thanks

**Pull request merged**